### ISSUE TITLE
libcxx run export conflict with sysroot installation of libcxx with old version of clangdev

### ISSUE THREAD

**Issue by SylvainCorlay** [conda-forge/clangdev-feedstock#188]
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

Some packages such as `cling` and `xeus-cling` have a runtime dependency on non-latest versions of `clangdev` (such as `clangdev 5`). Recent builds of clangdev 5 are compiled with more recent versions of clang, which have a strong run export on `libcxx`. As a consequence, recent builds of `clangdev 5` have a runtime dependency on `libcxx`, which is also installed in the SDK version shipped with XCode. 

Having libcxx in both the conda prefix and the sysroot poses some issues for old versions of clang, because libcxx uses the `include_next` statement to import files with same name in the hierarchy of include search directories, like [here](https://github.com/llvm-mirror/libcxx/blob/78d6a7767ed57b501/include/stdlib.h#L20). The duplication causes the `include_next` statement of the conda prefix to go to the sysroot's libcxx instead of the sysroot base directory.

This issue was resolved in recent versions of clang by https://reviews.llvm.org/D89001 ("Don't look into <sysroot> for C++ headers if they are found alongside the toolchain").

A possible solution would be to remove the search for libcxx in a patch for recent builds of clangdev 5 and other impacted versions, since they ship libcxx as a dependency anyways.

cc @isuruf @chrisburr FYI

### Installed packages

```shell
N/A
```

### Environment info

```shell
N/A
```

**Comment by JohanMabille** [comment]
After further investigations, it looks like the patch of clangdev 5 is required but not sufficient. We also need to patch recent versions of libcxx which use C++20 only features without preprocessor guards.

**Comment by SylvainCorlay** [comment]
clang 9 and therefore cling 0.9 are also impacted.

**Comment by AngryMaciek** [comment]
I came here from https://github.com/jupyter-xeus/xeus-cling/issues/403
Is there any progress on fixing this issue?

**Comment by h-vetinari** [comment]
Well, it sounds to me like the solution is simply to build clang 5/9 with old compilers, so that you don't get the run-export on new libcxx?

**Comment by AngryMaciek** [comment]
Thats a workaround, not a clean solution; How about:

> A possible solution would be to remove the search for libcxx in a patch for recent builds of clangdev 5 and other impacted versions, since they ship libcxx as a dependency anyways.

**Comment by h-vetinari** [comment]
> Thats a workaround, not a clean solution

How so? New compilers need newer support libraries, including libcxx. If you need older support libs, user older compilers. I mean, it _might_ work, but why risk that?

> Plus also: [#188 (comment)](https://github.com/conda-forge/clangdev-feedstock/issues/188#issuecomment-1290256280)

That's kind of my point -- we're not going to do:
> We also need to patch recent versions of libcxx which use C++20 only features without preprocessor guards.

That's completely outside the bounds of what's tested upstream, and puts the burden on us to ensure our newly-patched frankenstein actually doesn't break in creative ways. That's hardly a cleaner solution than using an old compiler _to build an ancient compiler_.

**Comment by AngryMaciek** [comment]
Fair enough, I come from `xeus-cling` thread mentioned above; in there @SylvainCorlay wrote he is "working on a cleaner solution" (https://github.com/jupyter-xeus/xeus-cling/issues/403#issuecomment-1288622579), so I am checking up.

**Comment by h-vetinari** [comment]
I mean, if there are patches that are only very light and obviously™ correct, than we can consider it (and I may be overruled by other maintainers, not saying I have any sort of final word). But I don't see the case to be honest. I'm happy to help keep the old compilers working, but that to me includes a period-appropriate dependency setup (incl. compilers used to build). Clang 5.0 is over 5 years [old](https://github.com/llvm/llvm-project/releases/tag/llvmorg-5.0.0) which is an eternity in this space.

I don't know the constraints of xeus, but moving beyond such old clang versions is hopefully on the roadmap already.

**Comment by h-vetinari** [comment]
Xref https://github.com/root-project/cling/issues/475

**Comment by sifatraquib** [comment]
any update?

**Comment by h-vetinari** [comment]
Someone needs to come up with a PR (to [libcxx](https://github.com/conda-forge/libcxx-feedstock), to make it possible for use with clang 5). Depending on how invasive it is, we'll see if this is achievable.

> Recent builds of clangdev 5 and 9 are compiled with more recent versions of clang, which have a strong run export on `libcxx`

Or you just try to build clangdev 5/9 here with older compilers, to avoid the problematic run-export in the first place.

I don't know the constraints w.r.t. cling well enough to do this myself, but I'll help with review etc. once something's there.

**Comment by h-vetinari** [comment]
Hey all

With https://github.com/conda-forge/libcxx-testing-feedstock/pull/3 resp. https://github.com/conda-forge/libcxx-feedstock/pull/127 we now have a way to test our libcxx against older clang builds (currently going back to 11, as that was what our lower bound for testing was before that PR).

That means libcxx 16 in conda-forge currently "supports"[^1] at least clang 11-16, which is already 3 versions more than upstream [policy](https://releases.llvm.org/16.0.0/projects/libcxx/docs/index.html#platform-and-compiler-support) (14, 15 & 16).

To get an idea of the degree of the breakage for older clangs, I opened https://github.com/conda-forge/libcxx-testing-feedstock/pull/5. It looks like clang 10 still passes, but from clang 9 on down, we have the problem with `is_trivially_constructible`.

[^1]: doesn't spectacularly blow up
